### ISSUE TITLE
config_path doesn't work with bash process substitution

### ISSUE THREAD

**Issue by directionless** [osquery/osquery#6040]
# Bug report

I'm not totally sure I'm testing this correctly, but I noticed that I can't use process substitution for the `config_path`. This feels like a bug.

```
# simple config file
$ cat /tmp/no_kernel.conf 
{ "options": { "disable_tables": "kernel_info"}}

# works when passed as a plain file
$ osqueryd -S --config_path /tmp/no_kernel.conf 'select version from kernel_info;'
Error: no such table: kernel_info

# unexpected when using process substitution
$ osqueryd -S --config_path <(cat /tmp/no_kernel.conf) 'select version from kernel_info;'
+---------+
| version |
+---------+
| 18.7.0  |
+---------+
```

### What operating system and version are you using?
```
 version = 10.14.6
   build = 18G1012
platform = darwin
```
### What version of osquery are you using?
```
version = 4.0.2
```

**Comment by sharvilshah** [comment]
I am not quite sure that this is a bug.

`config_path` semantics feels like it should be a path to a regular file.

We specifically check whether `config_path` is a regular file here:

https://github.com/osquery/osquery/blob/master/plugins/config/filesystem_config.cpp#L48-L50

Since substitution generates a file descriptor with an anonymous pipe `osqueryd` will error out.  We can see this by passing `--verbose` to `osqueryd`

```bash
➜  ~ cat /tmp/no_kernel.conf
{ "options": { "disable_tables": "kernel_info"}}

➜  ~ echo <(cat /tmp/no_kernel.conf)
/dev/fd/11

➜  ~ ls -lah /dev/fd/11
ls: /dev/fd/11: Bad file descriptor

➜  ~ ls -lah <(cat /tmp/no_kernel.conf)
prw-rw----  0 sbs  staff    49B Nov 14 14:58 /dev/fd/11

➜  osquery git:(master) ✗ ./build/osquery/osqueryd -S --config_path <(cat /tmp/no_kernel.conf) --verbose
I1114 14:59:42.315436 2881536896 init.cpp:693] Error reading config: config file does not exist: /dev/fd/11
Using a virtual database. Need help, type '.help'
osquery>
```

**Comment by directionless** [comment]
Thanks for digging that up! I wonder why we check for it being a regular file.

This appears to have been introduced in https://github.com/osquery/osquery/pull/950 as a mechanism to prevent people from passing in non-json, and crashing osquery. I'd say it was probably wrong -- if osquery can read it, it shouldn't matter if it's a regular file or not. 

Now that you point out `--verbose` has the error, I also see it. That feels like a simple fix. I think failure to read the config should be a warning, not a verbose log. But I can PR that for discussion.

As an aside, I don't think the `ls -lah /dev/fd/11` would work. The FD doesn't exist outside the process that got it.

**Comment by directionless** [comment]
That the shell lots that as INFO, not WARNING comes from https://github.com/osquery/osquery/pull/1510 in response to https://github.com/osquery/osquery/issues/1504 

Right now, that feels off to me. If someone specified a `config_path` not being able to read it is clearly an error. Not surfacing that creates confusion. (This ticket, some slack threads...) But since it was intentional, I want to think a little harder.

**Comment by sharvilshah** [comment]
> I'd say it was probably wrong -- if osquery can read it, it shouldn't matter if it's a regular file or not.

I think it's okay to have a clear and a definite boundary here, that osquery's filesystem config plugin will only read regular files. I would argue that it simplifies a lot of the code and maintaining that code, since operations and various access flags on a regular file are more widely understood. I am sure that Windows handles pipes/special files differently than POSIX environments. I am of the camp that the less code to maintain the better it is :)

> As an aside, I don't think the `ls -lah /dev/fd/11` would work. The FD doesn't exist outside the process that got it.

Yes, I quickly realized that :)
I do correct that in the next line.

> I suspect the underlying bug is that the config stuff doesn't differentiate between can't read the default config file, and can't read a user specified config file.

It's been a while since I have played around in that part of the codebase, so I don't know from the top off my head.

**Comment by sharvilshah** [comment]
> Right now, that feels off to me. If someone specified a `config_path` not being able to read it is clearly an error. Not surfacing that creates confusion. (This ticket, some slack threads...)

I do agree, and we could be doing a better job of surfacing the error.

**Comment by directionless** [comment]
> I think it's okay to have a clear and a definite boundary here, that osquery's filesystem config plugin will only read regular files. I would argue that it simplifies a lot of the code and maintaining that

I agree we should not maintain extra code, but I think we shouldn't head off what will probably work. I think we should try to read the file. If it works, great. If not, surface the error. Insisting on a regular file feels oddly limiting to some deployments, and since we can just try to read it, it's extra work.

**Comment by theopolis** [comment]
Yeap, this seems like a UX we can improve.

**Comment by directionless** [comment]
Thanks for the detailed report and clear reproduction steps. This issue has been open for quite a long time without activity. The behavior you're seeing is likely because bash process substitution creates a temporary pipe via `/dev/fd/`, and osquery's config reading expects a real file path (which it may need to re-read or watch for changes). A simple workaround is to write to a temp file first or use the plain file path as shown in your working example.

Since this is a fairly niche edge case and hasn't seen activity in many years, I'm going to close this out. If this is still relevant to you or others, please feel free to reopen — and contributions are always welcome if someone wants to add support for reading config from stdin or pipes. osquery is a volunteer project and we appreciate the community's help!
